File: Lucy/2025/5-W02/2300_Successful_Pairs_of_Spells_and_Potions.ts
function successfulPairs(spells: number[], potions: number[], success: number): number[] {
  const results: number[] = [];
  potions.sort((a, b) => a - b);

  for (const spell of spells) {
    let start = 0;
    let end = potions.length - 1;
    let answer = potions.length;
    while (start <= end) {
      const mid = Math.floor((start + end) / 2);

      if (potions[mid] >= success / spell) {
        answer = mid;
        end = mid - 1;
      } else {
        start = mid + 1;
      }
    }

    results.push(potions.length - answer);
  }

  return results;
}

File: Meli/LeetCode 75/2300. Successful Pairs of Spells and Potions.py
import bisect
from typing import List

class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        potions.sort()
        n = len(potions)
        answer = []
        for i in spells:
            answer.append(n-bisect.bisect_left(potions,success/i))
        return answer